feat: LEAP-1440: Allow to attach a region to a new comment

Gondragos opened the pull request:
This PR provides an ability to link a new comment to a region.

Other changes: 
- fix displaing regions without feature flag enabled https://github.com/HumanSignal/label-studio/pull/6400/files#diff-f4db8bd107bcd19f6cfa67c1b88e801a4a5608c73b12b2bae96fea65b34bec08R168


### PR fulfills these requirements
- [x] Commit message(s) and PR title follows the format `[fix|feat|ci|chore|doc]: TICKET-ID: Short description of change made` ex. `fix: DEV-XXXX: Removed inconsistent code usage causing intermittent errors`
- [ ] Tests for the changes have been added/updated (for bug fixes/features)
- [ ] Docs have been added/updated (for bug fixes/features)
- [x] Best efforts were made to ensure docs/code are concise and coherent (checked for spelling/grammatical errors, commented out code, debug logs etc.)
- [x] Self-reviewed and ran all changes on a local instance (for bug fixes/features)



#### Change has impacts in these area(s)
- [ ] Product design
- [ ] Backend (Database)
- [ ] Backend (API)
- [x] Frontend

Describe the reason for change
It's a part of implementing linking comment to region functionality.

#### What feature flags were used to cover this change?
`fflag_feat_all_leap_1430_per_field_comments_100924_short`



### Does this PR introduce a breaking change?
_(check only one)_
- [ ] Yes, and covered entirely by feature flag(s)
- [ ] Yes, and covered partially by feature flag(s)
- [x] No
- [ ] Not sure (briefly explain the situation below)



### What level of testing was included in the change?
_(check all that apply)_
- [ ] e2e
- [ ] integration
- [ ] unit



### Which logical domain(s) does this change affect?
`NewComment`

Gondragos committed:
Add Linking Modes for Annotations

Introduce `CommentMode` and `RelationMode` to handle linking annotations with regions. Also, refactor parts of the code to unify the method of linking actions under `LinkingModes` and replace deprecated `relationMode` checks.

Gondragos committed:
Add integration tests for linking modes to comments

Gondragos committed:
Refactor CommentMode and RelationMode to use volatile state.

Moved the `comment` and `firstRegion` variables to volatile state within the MobX-state-tree models for `CommentMode` and `RelationMode`. This change provides access to payload information of modes in case we need it.

Commit:
Fix warning message

Co-authored-by: hlomzik <hlomzik@gmail.com>

Commit:
Fix comment about deprecated method

Co-authored-by: hlomzik <hlomzik@gmail.com>

Commit:
Fix comment about deprecated method

Co-authored-by: hlomzik <hlomzik@gmail.com>

Gondragos committed:
Rename misnamed 'to_coments' to 'to_comments'

Gondragos committed:
Change cursor constants for linking mode

Updated all instances of `Constants.RELATION_MODE_CURSOR` to `Constants.LINKING_MODE_CURSOR` across various files. This ensures consistency in naming and improves code readability for linking mode features.

Gondragos committed:
Refactor LinkingModes model

Gondragos committed:
Fix linkingModes initing

Gondragos committed:
Update comment handling and linking UI

Refactored comments feature to support per-field comments through feature flags, added linking functionality between comments and regions, and updated UI with new icons and styles.

Gondragos committed:
Refactor comment form component structure

Reorganize the comment form components by separating old and new versions. Adjust CSS class names to avoid conflicts and import paths for clarity.

Gondragos committed:
Fix typo in CommentForm.scss

Gondragos commented on web/libs/editor/src/components/Comments/Comment/CommentForm.scss:
```diff
@@ -0,0 +1,109 @@
+    border: 1px solid var(--sand_300);
+    border-radius: 4px;
+    border-top: none;
+    bordet-top-left-radius: 0;
```
Thanks! I have updated the mentioned property name.

hlomzik commented on web/libs/editor/src/components/SidePanels/OutlinerPanel/OutlinerTree.tsx:
```diff
@@ -199,44 +199,47 @@ const OutlinerInnerTreeComponent: FC<OutlinerInnerTreeProps> = observer(({ regio
   );
 });
 
+export const displayRegionLabel = (item: any) => {
```
make this a proper react component maybe? it will look better as an exported resource + won't be worse to use it as component in this file as well. and we actually should have something similar in other places.

Gondragos replied:
Done, now it's a component.

Gondragos commented:
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/10942180936)
> Successfully merged:  create mode 100644 web/libs/editor/tests/integration/e2e/control_tags/classification/textarea.cy.ts

Gondragos committed:
Merge branch 'develop' into 'fb-leap-1440/link-new-comment'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/10942180936

Gondragos committed:
Fix FF_OUTLINER_OPTIM integration

Gondragos committed:
Refactor display of region labels

Replace `displayRegionLabel` with `RegionLabel` component in OutlinerTree and LinkState components.

Gondragos committed:
Remove unused IconWarning and Tooltip imports

Gondragos committed:
stylelint fixes

Gondragos committed:
biome fixes

Gondragos committed:
Missed file

bmartel commented on web/libs/editor/src/components/Comments/Comment/CommentForm.scss:
```diff
@@ -0,0 +1,110 @@
+.comment-form-enhanced {
+  $block: &;
```
```
  $block: &;
```

What does this do?

bmartel replied:
Ahh I found the usages later below. I felt like this layer of indirection makes it hard to understand what the requirement was. Maybe we just write the selectors without the SCSS vars string concatenation, as it is easier to read with locality of context for future readers including our design team. I will say that I do appreciate the programatic means in which you are thinking about it though 🙂.

hlomzik replied:
can we just convert this to css module? then it won't require those playing with nesting

Gondragos replied:
@hlomzik How to do it in the current state of everything?

bmartel commented on web/libs/editor/src/components/Comments/Comment/CommentForm.scss:
```diff
@@ -0,0 +1,110 @@
+  }
+
+  &_inline {
+    & #{$block}__actions {
```
```suggestion
    & .comment-form-enhanced__actions {
```
I don't think we need to be that DRY, it makes this really hard to understand the context.

hlomzik replied:
also it can be simply rewritten as `&_inline &__actions`. definitely no need in deep nesting and confusing var

Gondragos replied:
For me both options are worse than it is right now in terms of readability, cause `block` has semantics and `&` doesn't. And the `block` retains advantages of the BEM approach. But as you wish.

bmartel commented on web/libs/editor/src/stores/Comment/Anchor.js:
```diff
@@ -0,0 +1,25 @@
+
+export const Anchor = types
+  .model({
+    // id: types.optional(types.identifier, guidGenerator),
```
```
    // id: types.optional(types.identifier, guidGenerator),
```
Is this commented out to note that it is a type from another Model/Mixin?

Gondragos replied:
Oh, no, sorry, I just was not sure that I can skip defining `id` here. I'll remove it 😅

bmartel commented on web/libs/editor/src/components/SidePanels/OutlinerPanel/OutlinerTree.tsx:
```diff
@@ -206,37 +207,7 @@ const useDataTree = ({ regions, rootClass, footer }: any) => {
-        return item.value;
-      }
-    })();
+    const label = <RegionLabel item={item} />;
```
```
    const label = <RegionLabel item={item} />;
```
❤️

Gondragos replied:
That was Andrew's Idea 😊 https://github.com/HumanSignal/label-studio/pull/6400#discussion_r1765078142

bmartel commented on web/libs/editor/src/components/Comments/Comment/LinkState.tsx:
```diff
@@ -0,0 +1,68 @@
+  item: any;
+  onUnlink?: (item: any) => void;
+};
+const LinkedRegion: FC<LinkedRegionProps> = observer(({ item, onUnlink }) => {
```
```suggestion

const LinkedRegion: FC<LinkedRegionProps> = observer(({ item, onUnlink }) => {
```

bmartel commented on web/libs/editor/src/components/Comments/Comment/LinkState.tsx:
```diff
@@ -0,0 +1,68 @@
+      {mod?.display && <LinkedRegion item={region} onUnlink={onUnlink} />}
+    </Block>
+  );
+};
```
```suggestion
};

```

bmartel commented on web/libs/editor/src/components/Comments/Comment/LinkState.tsx:
```diff
@@ -0,0 +1,68 @@
+  region: any;
+  onUnlink?: (region: any) => void;
+};
+export const LinkState: FC<LinkStateProps> = ({ linking, region, onUnlink }) => {
```
```suggestion

export const LinkState: FC<LinkStateProps> = ({ linking, region, onUnlink }) => {
```

Commit:
Add blank lines for readability

Co-authored-by: bmartel <brandonmartel@gmail.com>

Commit:
Add blank line at eof

Co-authored-by: bmartel <brandonmartel@gmail.com>

Commit:
Add blank lines for readability

Co-authored-by: bmartel <brandonmartel@gmail.com>

Gondragos committed:
Remove block variable from scss

Gondragos committed:
Clear commented code

Gondragos committed:
stylelint fixes

bmartel reviewed (approved):
Really cool, overall a nice implementation and very interesting feature overall 👏
